### ISSUE TITLE
high quality figure

### ISSUE THREAD

**Issue by eguarnera** [3dmol/3Dmol.js#269]
Hi there,

Is it possible within py3dmol to save to high resolution figures, such as pdf, eps, tifff…

Best,
Enrico

**Comment by dkoes** [comment]
You can do this programatically use the native methods of the canvas object.  The resulting image will be the same size as the canvas. For example, the following code make right-clicking on the 3Dmol viewer return an image:
```
          $('canvas').bind('contextmenu',function(e) {
              var dt = e.currentTarget.toDataURL('image/png');
              window.location = dt;
          });
```
Would you prefer a simpler interface? What are your suggestions?

**Comment by eguarnera** [comment]
How can I use your code in py3dmol? This is the code that I use to visualize a local pdb in ipython, it would be great to have a command to simply save the image on screen...

cut=1
pdbdata=open('outfile.pdb', 'r').read()
view=py3Dmol.view()
view.addModel(pdbdata,'pdb')
view.zoomTo()
view.setBackgroundColor('white')
view.addSurface(py3Dmol.SAS,{'opacity':1,'colorscheme':{'prop':'b','gradient':'rwb','min':-cut,'max':cut}})
view

**Comment by dkoes** [comment]
Integration with py3Dmol requires modifying that package.  Which I have (and published the changes on PyPI).  view.png() will now embed a png of the current state of the viewer in the notebook.  However, I suspect that this is not what you want since 1) the image data is not available in python (e.g. you can't save the image to a file), 2) if you use this call non-interactively, you get a white image since it snapshots the viewer before anything is rendered, and 3) since a save does a re-render, saved notebook have white images.

Basically, the only thing it's good for is you can manually extract the image from the notebook (e.g. right-click or drag).  Since the image is the same size of the viewer, this doesn't give you anything higher resolution than a screenshot though (but might be a bit more convenient).

Can you describe your desired use case more?

**Comment by eguarnera** [comment]
Hi,

thanks for updating py3Dmol to 0.5.3. I tried the command view.png() in this way

view.show()
figfile=open('fig.png','w')
view.png(figfile)

which gives the error

---> 17 view.png(figfile)
TypeError: png() takes exactly 1 argument (2 given)

I'm using py3Dmol interactively with ipython to visualize protein surfaces colored according to a certain model. So I'm primarily using py3Dmol for the visual inspection of my model. However, since the rendered images look very nice, I was considering to use them for the figures in the paper I'm writing. I tried to use the right-click to save the image but it simply doesn't work, that is right-click over the image doesn't open any menu.

**Comment by dkoes** [comment]
The png method doesn't take any arguments.  Like I said previously, we cannot import the image data into python and so can't save files directly.

If you are right-clicking on the 3D viewer, you won't get a menu.  If you right-click on the image generated by view.png() you will get a menu.

**Comment by eguarnera** [comment]
I see. I followed your instructions, I'm attaching two screenshots, not using and using view.png() respectively. When I use view.png()  I get the an error shown in figure. 

![3dmol1](https://cloud.githubusercontent.com/assets/8078280/22359834/bee32c3e-e485-11e6-9d32-e55974b6548e.jpeg)
![3dmol2](https://cloud.githubusercontent.com/assets/8078280/22359837/c8c08d14-e485-11e6-859b-6edbe5b644ce.jpeg)

**Comment by eguarnera** [comment]
What am I doing wrong, any suggestion? Thanks for you time!

Best,
Enrico

**Comment by dkoes** [comment]
Perhaps if you uploaded the notebook?

**Comment by eguarnera** [comment]
matplotlib inline
%config InlineBackend.figure_format='retina'

from pylab import *
import py3Dmol

cut=1
pdbdata=open('outfile.pdb', 'r').read()
view=py3Dmol.view()
view.addModel(pdbdata,'pdb')
view.zoomTo()
view.setBackgroundColor('white')
view.addSurface(py3Dmol.SAS,{'opacity':1,'colorscheme':{'prop':'b','gradient':'rwb','min':-cut,'max':cut}})
view.show()
# view.png()

**Comment by eguarnera** [comment]
below the pdb outfile.pdb

**Comment by eguarnera** [comment]
[outfilepdb.txt](https://github.com/3dmol/3Dmol.js/files/741008/outfilepdb.txt)

**Comment by dkoes** [comment]
No, I meant upload the actual .ipynb file.  Your code works fine for me.  I want to see what is different in your environment.

**Comment by eguarnera** [comment]
[test3Dmol.ipynb.zip](https://github.com/3dmol/3Dmol.js/files/741046/test3Dmol.ipynb.zip)

**Comment by dkoes** [comment]
I can't reproduce the problem or see any reason for it.  I've tried with Safari, Firefox and Chrome.  I'm using ipython 5.1 with notebook server 4.2.3.  What is your setup?

**Comment by eguarnera** [comment]
I have ipython 5.1.0 and notebook 4.3.1… I don't understand

**Comment by eguarnera** [comment]
notebook running on Firefox 50.1

**Comment by dkoes** [comment]
I upgraded to 4.3.1 and Firefox 51.0.1 - still no problem.  Have you try restarting your browser, restarting the ipython kernel, and/or creating a new notebook?

**Comment by eguarnera** [comment]
Ok, giving the command view.png() in a new cell works! Giving the command right after view.show() in the same cell gives me the error. But fine, thanks!

**Comment by dkoes** [comment]
Yes, the viewer has to be instantiated for the command to work.

**Comment by eguarnera** [comment]
Sorry, had misunderstood how it worked. Thanks for your patience and time!

**Comment by miRastic** [comment]
The below sample code worked for me:

          $('canvas').bind('contextmenu',function(e) {
              var dt = e.currentTarget.toDataURL('image/png');
              var a = document.createElement("a"); //Create <a>
              a.href = dt;
              a.download = "Image.png"; //File name Here
              a.click(); //Downloaded file
              a.remove();
          });
